### PR TITLE
Add Default Callback Name editor setting

### LINKED ISSUE #1023: Allow users to specify signal method receiver name format
**Describe the project you are working on:**
A game's UI

**Describe the problem or limitation you are having in your project:**
I'm currently reorganizing the user interface of a game and whenever I rename an instanced scene, or I change the "Edit Children" property of the instanced scene, I lose all my signal connections that live outside of those instanced scenes.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Instead of generating a string of type `_on_${node.name}_${node.signal_in_question}`, let users specify what string they want to use when connecting signals by using other property values of the node.

For example, by editing the project properties, users can then specify the signal format of: `_on_${node.name}_${node.class}_${node.signal_in_question}`

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Not entirely. Connecting signals through GDScript has the side effect of not showing up in the UI which is not what we want in our current project given the sheer amount of signals.

It's worth mentioning that this issue will come up when working on complex games -- refactoring nodes and code whilst keeping the UI the same will become a pain because users will have to retype or copy paste function names every single time they attempt to connect signals through the editor.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Godot's editor must be modified to add this feature.

### LINKED ISSUE #3346: Add Default Callback Name editor setting for Signal Connection Dialog
### Describe the project you are working on

A game.

### Describe the problem or limitation you are having in your project

By default, Signal Connection Dialog generates names that violate GDScript [naming conventions](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_styleguide.html#functions-and-variables).

And you have to rename this every time.

The setting I propose allows you to set, for example, the following formats:

```
_on_{node_name}_{signal_name}
_{node_name}_{signal_name}
_on_{signal_name}
_{signal_name}
On{NodeName}{SignalName}
On{SignalName}
```

You can even specify an empty string if you come up with a good method name every time and you don't need an editor's suggestion anyway.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add the "interface/editors/default_callback_name" editor setting, which allows you to specify the format of the default callback name in the Signal Connection Dialog.

Available substitutions: `{NodeName}`, `{nodeName}`, `{node_name}`, `{SignalName}`, `{signalName}`, `{signal_name}`.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

godotengine/godot#52997

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, this is the functionality of the editor.

### Is there a reason why this should be core and not an add-on in the asset library?

As far as I know, this cannot be implemented by a plugin (without hacks).

### PR DESCRIPTION
**Pull request by dalexeev**
Adds the "interface/editors/default_signal_callback_name" editor setting,
which allows you to specify the format of the default callback name
in the Signal Connection Dialog.

---

Available substitutions: `{NodeName}`, `{nodeName}`, `{node_name}`, `{SignalName}`, `{signalName}`, `{signal_name}`.

Closes godotengine/godot-proposals#1023
Closes godotengine/godot-proposals#3346.

### CODE REVIEW

**Comment by groud** [comment]
Is there a proposal attached to this PR ? I am not sure such customization is very useful.

**Comment by dalexeev** [comment]
See godotengine/godot-proposals#3346.

File: editor/connections_dialog.cpp [hunk 1]
```diff
@@ -712,38 +712,56 @@ bool ConnectionsDock::_is_item_signal(TreeItem &item) {
...
+
+	s = signal_name.capitalize().replace(" ", "");
+	subst["SignalName"] = s;
+	if (s.length()) {
+		s[0] = s.to_lower()[0];
+	}
+	subst["signalName"] = s;
+	subst["signal_name"] = signal_name.capitalize().replace(" ", "_").to_lower();
+
+	String dst_method = String(EDITOR_DEF("interface/editors/default_callback_name", "_on_{node_name}_{signal_name}")).format(subst);
```
Comment (KoBeWi) [inline editor/connections_dialog.cpp:757]: This should be `EDITOR_GET`. Defining something twice is a bad idea.

File: editor/connections_dialog.cpp [hunk 2]
```diff
@@ -712,38 +712,56 @@ bool ConnectionsDock::_is_item_signal(TreeItem &item) {
...
 	if (!dst_node || dst_node->get_script().is_null()) {
 		dst_node = _find_first_script(get_tree()->get_edited_scene_root(), get_tree()->get_edited_scene_root());
 	}
 
-	StringName dst_method = "_on_" + midname + "_" + signal;
+	Dictionary subst;
+
+	String s = node_name.capitalize().replace(" ", "");
+	subst["NodeName"] = s;
+	if (s.length()) {
```
Comment (KoBeWi) [inline editor/connections_dialog.cpp:743]:
```suggestion
	if (!s.is_empty()) {
```

File: editor/connections_dialog.cpp [hunk 3]
```diff
@@ -712,38 +712,56 @@ bool ConnectionsDock::_is_item_signal(TreeItem &item) {
...
+		s[0] = s.to_lower()[0];
+	}
+	subst["nodeName"] = s;
+	subst["node_name"] = node_name.capitalize().replace(" ", "_").to_lower();
+
+	s = signal_name.capitalize().replace(" ", "");
+	subst["SignalName"] = s;
+	if (s.length()) {
```
Comment (KoBeWi) [inline editor/connections_dialog.cpp:751]:
```suggestion
	if (!s.is_empty()) {
```

**Review by KoBeWi — APPROVED** [review]
Looks good, my only concern (aside from the comments I left) is that it's possible to put an invalid identifier as a default callback. Not sure how to safeguard against that though, other than extending the validation from `node_name` to the whole callback. Also the setting name should be `default_signal_callback_name`, to be more clear.

Though I'd wait with bigger changes until the PR is approved as desired. Personally I almost never use the default name, so I'd be in favor of this feature, but the proposal itself didn't gain much interest.

**Comment by dalexeev** [comment]
See also the [comment](https://github.com/godotengine/godot-proposals/issues/3553#issuecomment-971806667).

**Comment by akien-mga** [comment]
Reviewed and approved in PR review meeting, the feature makes sense for users who want to customize this to their own style guides.

Needs a rebase (there's been a PR merged recently which also did style fixes to this code).

**Review by akien-mga — APPROVED** [review]

**Comment by dalexeev** [comment]
I think this should be a project setting, not an editor setting, as different projects may have different code style conventions. And in general, we should unify those editor settings that can be overridden in the project settings.

**Review by akien-mga — APPROVED** [review]
It's ready so let's merge, but for future PRs please make sure to keep cosmetic changes separate from functional changes. It's very difficult to review the actual logic changes in this commit in the midst of all the renames.

**Comment by akien-mga** [comment]
Thanks!

> I think this should be a project setting, not an editor setting, as different projects may have different code style conventions. And in general, we should unify those editor settings that can be overridden in the project settings.

That's matter for a proposal discussion IMO, and affects a lot more than just this setting (e.g. whether to use tabs or spaces for indentation, etc.).

**Comment by dalexeev** [comment]
> It's ready so let's merge, but for future PRs please make sure to keep cosmetic changes separate from functional changes. It's very difficult to review the actual logic changes in this commit in the midst of all the renames.

Okay. Initially I made two commits, but before the rebase I squash them into one (according to the [docs](https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html)).

**Comment by akien-mga** [comment]
We often ask to squash commits when they reflect a single semantic change split over multiple check-ins (e.g. you add a method, then in another commit you bind it, then in a third commit you add docs - all this is a single semantic change -- new API -- which should be a single commit). But sometimes we do want to keep multiple commits when they do separate things which are valuable to keep separate, such as splitting a big cosmetic overhaul from actually implementing a feature in its midst.

**Pull request merged**